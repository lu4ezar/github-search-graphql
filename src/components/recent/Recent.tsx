import React from "react";
import Octicon, { ChevronUp, ChevronDown } from "@primer/octicons-react";
import {
  HistoryItem,
  RecentDiv,
  Panel,
  Arrow,
  Button,
  CollapsableListDiv
} from "./styled";

type RecentProps = {
  searchHistory: string[];
  getSearchStringFromHistory: (str: string) => void;
};

const Recent = ({ searchHistory, getSearchStringFromHistory }: RecentProps) => {
  const [isCollapsed, changeCollapseState] = React.useState(true);
  const list: React.ReactNode[] | null = searchHistory.length
    ? searchHistory.map(
        (searchWord, index: number): React.ReactNode => (
          <HistoryItem
            title={`search for "${searchWord}" again`}
            key={searchWord}
            onClick={() => getSearchStringFromHistory(searchWord)}
            onKeyDown={() => getSearchStringFromHistory(searchWord)}
            role="button"
            tabIndex={index}
          >
            {searchWord}
          </HistoryItem>
        )
      )
    : null;
  const onClick = () =>
    isCollapsed ? changeCollapseState(false) : changeCollapseState(true);
  const currentIcon = isCollapsed ? ChevronDown : ChevronUp;
  return (
    <RecentDiv>
      <Panel>
        <Button type="button" onClick={onClick}>
          Search History
          <Arrow className={!searchHistory.length ? "invisible" : ""}>
            <Octicon icon={currentIcon} size="medium" />
          </Arrow>
        </Button>
      </Panel>
      <CollapsableListDiv className={isCollapsed ? "collapsed" : ""}>
        {list}
      </CollapsableListDiv>
    </RecentDiv>
  );
};

export default Recent;
