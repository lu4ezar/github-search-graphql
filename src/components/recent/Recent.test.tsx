/* eslint-disable react/jsx-props-no-spreading */
import React from "react";
import { render, fireEvent, cleanup } from "@testing-library/react";
import Recent from "./Recent";
import "jest-styled-components";

const recentProps = {
  searchHistory: ["abc", "def", "ghi"],
  getSearchStringFromHistory: jest.fn()
};

const recentPropsNoData = {
  searchHistory: [],
  getSearchStringFromHistory: jest.fn()
};

const renderComponent = () => ({
  ...render(<Recent {...recentProps} />)
});
const renderComponentNoData = () => ({
  ...render(<Recent {...recentPropsNoData} />)
});

afterEach(cleanup);

it("matches the snapshot", () => {
  const { container } = renderComponent();
  expect(container).toMatchSnapshot();
});

it("renders list of recent searches", () => {
  const { getAllByTitle } = renderComponent();
  const listItems = getAllByTitle(/search for/i);
  expect(listItems).toHaveLength(3);
});

it("calls getSearchStringFromHistory prop function on List Item click", () => {
  const { getByText } = renderComponent();
  const listItem = getByText("abc");
  fireEvent.click(listItem);
  expect(recentProps.getSearchStringFromHistory).toBeCalledTimes(1);
  fireEvent.keyDown(listItem);
  expect(recentProps.getSearchStringFromHistory).toBeCalledTimes(2);
  fireEvent.keyDown(listItem);
  expect(recentProps.getSearchStringFromHistory).toBeCalledTimes(3);
  fireEvent.keyDown(listItem);
  expect(recentProps.getSearchStringFromHistory).toBeCalledTimes(4);
});

it("calls getSearchStringFromHistory prop function on List Item key press", () => {
  const { getByText } = renderComponent();
  const listItem = getByText("abc");
  fireEvent.keyDown(listItem);
  expect(recentProps.getSearchStringFromHistory).toBeCalled();
});

it("collapses list by clicking on its header", () => {
  const { getByText, getByRole } = renderComponent();
  const header = getByText("Search History");
  const list = getByRole("list");
  expect(list.classList.contains("collapsed")).toBe(true);
  fireEvent.click(header);
  expect(list.classList.contains("collapsed")).toBe(false);
  fireEvent.click(header);
  expect(list.classList.contains("collapsed")).toBe(true);
});

it("does not show 'Collapse/Expand' button on desktop", () => {
  const { getByTitle } = renderComponent();
  const icon = getByTitle(/expand/i);
  // @ts-ignore
  expect(icon).toHaveStyleRule({
    media: `(min-width: 1024px) {
      display: none;
    }`
  });
});

it("does not show 'Collapse/Expand' button if list is empty", () => {
  const { queryByTitle } = renderComponentNoData();
  const icon = queryByTitle(/expand/i);
  expect(icon).toBe(null);
});

it("shows 'Collapse/Expand' button if list is not empty", () => {
  const { getByTitle } = renderComponent();
  const icon = getByTitle(/expand/i);
  expect(icon.classList.contains("invisible")).toBe(false);
});
