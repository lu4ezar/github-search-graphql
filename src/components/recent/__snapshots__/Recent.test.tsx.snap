// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
.c0 {
  grid-area: aside_recent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  background-color: transparent;
  color: #262323;
  text-transform: uppercase;
  font-weight: bold;
  font-size: 1rem;
  cursor: pointer;
  border: none;
  outline: none;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  position: relative;
  right: -2rem;
}

.c4 {
  margin: 0;
  padding: 0;
  overflow: hidden;
  max-height: 100%;
  -webkit-transition: max-height 0.2s;
  transition: max-height 0.2s;
}

.c4.collapsed {
  max-height: 0px;
  -webkit-transition: max-height 0.2s;
  transition: max-height 0.2s;
}

.c5 {
  list-style-type: none;
  text-transform: lowercase;
  font-style: italic;
  cursor: pointer;
  margin: 1rem;
  font-weight: normal;
  -webkit-transition: background 0.4s;
  transition: background 0.4s;
  border-radius: 2px;
  padding: 0.2rem;
  text-align: center;
}

.c5:hover {
  background: #e6e6e6;
}

@media (min-width:1024px) {
  .c2 {
    cursor: auto;
  }
}

@media (min-width:1024px) {
  .c3 {
    display: none;
  }
}

@media (min-width:1024px) {
  .c4.collapsed {
    max-height: 100%;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="c2"
        type="button"
      >
        Search History
        <span
          class="c3"
          title="expand/collapse"
        >
          <svg
            aria-hidden="true"
            class="octicon"
            height="32"
            role="img"
            style="display: inline-block; fill: currentColor; user-select: none; vertical-align: text-bottom;"
            viewBox="0 0 10 16"
            width="20"
          >
            <path
              d="M5 11L0 6l1.5-1.5L5 8.25 8.5 4.5 10 6l-5 5z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
    <ul
      class="c4 collapsed"
    >
      <li
        class="c5"
        role="button"
        tabindex="0"
        title="search for \\"abc\\" again"
      >
        abc
      </li>
      <li
        class="c5"
        role="button"
        tabindex="1"
        title="search for \\"def\\" again"
      >
        def
      </li>
      <li
        class="c5"
        role="button"
        tabindex="2"
        title="search for \\"ghi\\" again"
      >
        ghi
      </li>
    </ul>
  </div>
</div>
`;
